src/tests/utests/xi_utest_data_desc.c

```diff
@@ -244,6 +244,259 @@ XI_TT_TESTCASE_WITH_SETUP(
+        const size_t memory_limit_size = XI_MEMORY_LIMITER_SYSTEM_MEMORY_LIMIT +
+                                         allocation_space;
+
+        /* assumptions */
+        tt_assert( XI_MEMORY_LIMITER_SYSTEM_MEMORY_LIMIT > allocation_footprint );
+
+        xi_state_t memory_set_result = xi_set_maximum_heap_usage( memory_limit_size );
+        tt_assert( memory_set_result == XI_STATE_OK );
+
+        XI_ALLOC_BUFFER_AT( unsigned char, origin_buffer, num_buffer_elements, ret_state );
+
+        /* test behavior */
+        data_desc = xi_make_desc_from_buffer_copy( origin_buffer, num_buffer_elements );
+        tt_ptr_op( data_desc, ==, NULL );
+        
+        goto end;
+
+    err_handling:
+        tt_fail();
+    end:
+        if( data_desc != NULL )
+        {
+            xi_free_desc( &data_desc );
+        }
+        XI_SAFE_FREE( origin_buffer );
+        xi_memory_limiter_set_limit( XI_MEMORY_LIMITER_SYSTEM_MEMORY_LIMIT +
+                                     XI_MEMORY_LIMITER_APPLICATION_MEMORY_LIMIT );
+
+    } )
+
+
+    XI_TT_TESTCASE(
+    utest__xi_make_desc_from_buffer_copy__valid_data__not_enough_memory_for_descriptor,
+    {
+        /* size values */
+        size_t num_buffer_elements   = 700;
+        xi_state_t ret_state         = XI_STATE_OK;
+        unsigned char* origin_buffer = NULL;
+        xi_data_desc_t* data_desc    = NULL;
+        
+        const size_t buffer_size_in_bytes = num_buffer_elements * sizeof( unsigned char );
+        const size_t allocation_space   = buffer_size_in_bytes +
+                                          sizeof( xi_data_desc_t ) +
+                                          sizeof( xi_data_desc_t ) +
+                                          2*sizeof( xi_memory_limiter_entry_t );
+        const size_t allocation_footprint = buffer_size_in_bytes +
+                                            sizeof( xi_memory_limiter_entry_t );
+        const size_t memory_limit_size = XI_MEMORY_LIMITER_SYSTEM_MEMORY_LIMIT +
+                                         allocation_space;
+
+        /* assumptions */
+        tt_assert( XI_MEMORY_LIMITER_SYSTEM_MEMORY_LIMIT > allocation_footprint );
+
+        xi_state_t memory_set_result = xi_set_maximum_heap_usage( memory_limit_size );
+        tt_assert( memory_set_result == XI_STATE_OK );
+
+        XI_ALLOC_BUFFER_AT( unsigned char, origin_buffer, num_buffer_elements, ret_state );
+
+        /* test behavior */
+        data_desc = xi_make_desc_from_buffer_copy( origin_buffer, num_buffer_elements );
+        tt_ptr_op( data_desc, ==, NULL );
+        
+        goto end;
+
+    err_handling:
+        tt_fail();
+    end:
+        if( data_desc != NULL )
+        {
+            xi_free_desc( &data_desc );
+        }
+        XI_SAFE_FREE( origin_buffer );
+        xi_memory_limiter_set_limit( XI_MEMORY_LIMITER_SYSTEM_MEMORY_LIMIT +
+                                     XI_MEMORY_LIMITER_APPLICATION_MEMORY_LIMIT );
+
+    } )
```

Comment:
```
441     XI_TT_TESTCASE(
442     utest__xi_make_desc_from_buffer_copy__valid_data__not_enough_memory_for_descriptor_atigyi,
443     {
444         /* size values */
445         const unsigned char origin_buffer[] = "Test data";
446         const size_t buffer_size_in_bytes = sizeof(origin_buffer);
447 
448         const size_t allocation_space   = sizeof( xi_data_desc_t ) +
449                                           sizeof( xi_memory_limiter_entry_t ) +
450                                           buffer_size_in_bytes +
451                                           sizeof( xi_memory_limiter_entry_t ) - 1;
452 
453         const size_t memory_limit_size = XI_MEMORY_LIMITER_SYSTEM_MEMORY_LIMIT +
454                                          allocation_space;
455 
456         xi_state_t memory_set_result = xi_set_maximum_heap_usage( memory_limit_size );
457         tt_assert( memory_set_result == XI_STATE_OK );
458 
459         /* test behavior */
460         xi_data_desc_t* data_desc = xi_make_desc_from_buffer_copy( origin_buffer, buffer_size_in_bytes );
461         tt_ptr_op( data_desc, ==, NULL );
462 
463     end:
464         xi_free_desc( &data_desc );
465         xi_memory_limiter_set_limit( XI_MEMORY_LIMITER_SYSTEM_MEMORY_LIMIT +
466                                      XI_MEMORY_LIMITER_APPLICATION_MEMORY_LIMIT );
467 
468     } )
```

Reply:
I would go with the implementation above. What it has in difference are:
- it uses an initialized array on the stack instead of using XI_ALLOC_BUFFER. This simplifies the test cases a bit by not needing the err_handling and simpler, bit clearer calculation of `allocation_space`.
- also I've removed the footprint check since I don't get what it is for. Maybe you will describe it for me.
- no need for guarding `xi_free_desc` it must handle NULL values.

Reply:
Let's discuss these suggestions and after please update all test cases to make them simpler.
Also could you please add a comment how the allocation_space is calculated.